### PR TITLE
storage/flash_map: add `get erase byte value API

### PR DESCRIPTION
**Pull request by nvlsianpu**
Added flash_area_erased_val() function for get value of erased
byte of memory which is under flash area.

~~Added flash_area_read_is_empty() function for reads data and check if
read data was erased.~~

These functions already exist in MCUBoot and zephyr dfu subsystem
which makes simultaneous usage of both impossible.

In the same time the mcuboot's function is made __weak for backward compatibility.
- [x] https://github.com/JuulLabs-OSS/mcuboot/pull/851
- [x] inherited by the zephyr-fork: https://github.com/zephyrproject-rtos/mcuboot/pull/38

### CODE REVIEW

**Review by utzig — APPROVED** [review]
LGTM

**Review by de-nordic — APPROVED** [review]

File: subsys/storage/flash_map/flash_map.c
Comment (utzig) [inline subsys/storage/flash_map/flash_map.c:288]: `erased_val` (with a `d`!)

File: subsys/storage/flash_map/flash_map.c
Comment (utzig) [inline subsys/storage/flash_map/flash_map.c:295]: erased_val (with a d!)

**Review by utzig — CHANGES_REQUESTED** [review]

**Review by utzig — APPROVED** [review]

**Comment by utzig** [comment]
I tested this on `frdm_k64f` with MCUboot and `smp_svr` (mcumgr) and everything works correctly. I was trying to test a board with erased to 0, the `b_l072z_lrwan1` in this case, but I am currently struggling to get it working; but at least I assured that MCUboot is receiving the correct erase value. I think you can remove the `[RFC]` and it would be nice to have this merged for next release.

File: subsys/storage/flash_map/flash_map.c
Comment (de-nordic) [inline subsys/storage/flash_map/flash_map.c:293]: I think that this function may return 1 (memory is erased) in case when `len` is given 0. `len == 0` is unusual situation, how should we handle it in this case?

Comment (nvlsianpu) [reply subsys/storage/flash_map/flash_map.c]: Let's be more precise on that, `len == 0` imposes returns of `-EINVAL`.

**Comment by nvlsianpu** [comment]
I changed `flash_area_read_is_empty()` that it returns zephyr like error-codes ("-Errno"), and treats an attempt to 0-length as an error as well.

File: include/storage/flash_map.h
Comment (pabigot) [inline include/storage/flash_map.h:256]: Should document the constraint on valid lengths here:

    len Number of bytes to read.  The value must be positive.

Though honestly I don't like the name of this function *or* its API, relative to what it does.  Since we're refactoring, can we correct the API: 

    int flash_area_try_read(const struct flash_area *fa, off_t off,
￼			     void *dst, size_t len, bool* is_erased);

?  Then we don't have a really non-standard behavior when `len == 0` that would make callers have to check the value.

I don't really like that name either; but I understand the role of this to be the same as a normal read, but to also indicate whether what was read was erased.

Maybe we shouldn't add this, but instead have:

    bool flash_area_erased(const struct flash_area *fa, const void *data, size_t len);

that returns `true` or `false`, with `false` if somebody passes `NULL` .

Yeah, I think that makes more sense.  It can even be a helper function that doesn't need to get implemented in each driver.

Comment (pabigot) [reply include/storage/flash_map.h:256]: Corrected badly chosen original name to "flash_area_*data*_is_erased".  It could also be "flash_area_buffer_is_erased".

Comment (utzig) [reply include/storage/flash_map.h:256]: Those names where not created here, they come from MCUboot's API requirements.

File: include/storage/flash_map.h
Comment (pabigot) [inline include/storage/flash_map.h:258]: Since this API is mandated by MCUboot we have to do what it specifies.   (That should be mentioned in the description of the function.)

Ah.  I see.  In that case, MCUboot should be specifying what the function return value should be under all conditions.  Instead of using a function in common code that requires every back-end to define and declare it (where every backend appears to be `cypress` and `zephyr` and `sim`.

MCUboot failed to be clear, and there are other ports that Zephyr's implementation must behave consistently with. MCUboot  `docs/PORTING.md` currently has:

```
/*< Reads len bytes from off, and checks if the read data is erased. Returns
    1 if empty (that is containing erased value), 0 if not-empty, and -1 on
    failure. */
int     flash_area_read_is_empty(const struct flash_area *fa, uint32_t off,
                     void *dst, uint32_t len);
```
While the text there could support `len == 0` returning zero, all three current back-ends (cypress, zephyr, and sim) return 1 in that case.

So this function needs to do so too.

**Review by pabigot — CHANGES_REQUESTED** [review]
As some of this API is defined by MCUboot we need to conform to its expectations.

We should also document that this API is specifically targeting MCUboot, and that MCUboot defines its behavior, so somebody doesn't come back later and try to fix it to be more sensible.

File: include/storage/flash_map.h
Comment (pabigot) [reply include/storage/flash_map.h:258]: Also the error return value is `-1`, nothing else.

Comment (de-nordic) [reply include/storage/flash_map.h:258]: > While the text there could support `len == 0` returning zero, all three current back-ends (cypress, zephyr, and sim) return 1 in that case.

Seems that they took `Returns 1 if empty` literally, buffer with `len == 0` of data is indeed empty.

Comment (de-nordic) [reply include/storage/flash_map.h:258]: > MCUboot failed to be clear, and there are other ports that Zephyr's implementation must behave consistently with. MCUboot `docs/PORTING.md` currently has:

Maybe we should propose changes to Mcuboot, make the documentation more clear and/or refactor it there with cooperation with other back-end providers.

Comment (utzig) [reply include/storage/flash_map.h:258]: > Maybe we should propose changes to Mcuboot, make the documentation more clear and/or refactor it there with cooperation with other back-end providers.

You would also have to submit a PR to Mynewt, since the API was originally generated there. I don't expect much friction, but I would defer merging it to other committers.

Comment (pabigot) [reply include/storage/flash_map.h:258]: I don't think it's worth trying to change upstream.  I'd rather seek us go ahead with `flash_area_erased_value()` which is independently useful, and leave `flash_area_read_is_empty()` as something that exists only to conform to MCUboot's existing API.  Ideally that function would be conditional on a Kconfig signalling MCUboot support.

Comment (nvlsianpu) [reply include/storage/flash_map.h:258]: Maybe we can change description of the return value in MCUBoot. MCUboottreads any negative value as an error.
Also name might be changed (not sure, @utzig will check that).
Will ensure consistence of behavior while `len==0`.

File: include/storage/flash_map.h
Comment (utzig) [reply include/storage/flash_map.h:258]: > Though honestly I don't like the name of this function _or_ its API, relative to what it does. Since we're refactoring, can we correct the API:
> 
> ```
> int flash_area_try_read(const struct flash_area *fa, off_t off, void *dst, size_t len, bool* is_erased);
> ```
> 

This function's name does not tell me that it's is checking for an erased block, although I can see by the last parameter's name...

> 
> ```
> bool flash_area_data_is_erased(const struct flash_area *fa, const void *data, size_t len);
> ```
> 
> that returns `true` or `false`, with `false` if somebody passes `NULL` .
> 
> Yeah, I think that makes more sense. It can even be a helper function that doesn't need to get implemented in each driver.

AFAIR the original idea was to have a way of read and check that it might be erased in a single read, so that users of the API don't need to manually to the erased checking. If I understand correctly this API would require two reads.

File: include/storage/flash_map.h
Comment (utzig) [reply include/storage/flash_map.h:260]: > Maybe we can change description of the return value in MCUBoot. MCUboottreads any negative value as an error.

Makes sense, send PR!

> Also name might be changed (not sure, @utzig will check that).

This is used only by MCUboot and fcb/fcb2 in Mynewt so should be easy to change when we settle on some better naming convention. Btw, the `fcb` in Zephyr won't work on devices with flash erased to 0 in its current form.

File: include/storage/flash_map.h
Comment (pabigot) [reply include/storage/flash_map.h:258]: No, that wasn't the intent.  Every implementation of `flash_area_read_is_empty` is basically:
* invoke the standard flash read
* iterate over the result looking for a non-erased value.

What I'm proposing with the badly named `flash_area_data_is_erased` is to make a function that does the second half, given the data that was read in the first half.  The only reason this is even in the flash area API is that we need an `fa` parameter is to find out what erase value we're checking against.  There's no I/O involved.

This would be generically useful, e.g. if you only want to check whether a certain sub-sequence of what you read was erased, so it makes a some sense to add it to Zephyr's API.

Then for applications that require `flash_area_data_is_erased` it's just a call to `flash_area_read` followed by a call to the function that checks it's all erased.

Comment (utzig) [reply include/storage/flash_map.h:258]: Ok, I like that.

**Comment by utzig** [comment]
This is currently missing Zephyr and Cypress backend updates, I will do it in a bit...

**Review by utzig — CHANGES_REQUESTED** [review]
Better wait for https://github.com/JuulLabs-OSS/mcuboot/pull/809

**Comment by utzig** [comment]
> Better wait for [JuulLabs-OSS/mcuboot#809](https://github.com/JuulLabs-OSS/mcuboot/pull/809)

Just merged it, you can move on with removal of functions from flash_map.

File: include/storage/flash_map.h
Comment (nvlsianpu) [reply include/storage/flash_map.h]: this function is not in scope anymore

File: include/storage/flash_map.h
Comment (nvlsianpu) [reply include/storage/flash_map.h]: this function is not in scope anymore

**Comment by nvlsianpu** [comment]
@utzig @pabigot please revisit.

**Review by pabigot — APPROVED** [review]
This seems fine, but note that we need to get JuulLabs-OSS/mcuboot#851 merged and the manifest updated to a real SHA before merging.  So it should be DNM until then.

**Review by utzig — APPROVED** [review]

**Comment by nvlsianpu** [comment]
mcuboot manifest SHA has been updated

**Pull request merged**
